draft-ietf-mls-protocol.md

```diff
@@ -1444,110 +1452,169 @@ To authenticate a message from a particular member, signatures are required.
 
-### Updating the Application Secret
+* node_index = IndexOf(V)
 
-The following rules apply to an Application Secret:
+Moreover, the secrets of the children of V are defined to be:
 
-- Senders MUST only use the Application Secret once and monotonically
-  increment the generation of their secret. This is important to provide
-  Forward Secrecy at the level of Application messages. An attacker getting
-  hold of a member specific Application Secret at generation [N+1] will not be
-  able to derive the member's Application Secret [N] nor the associated
-  AEAD key and nonce.
+~~~~
+astree_node_[IndexOf(V)]_secret
+        |
+        |
+        +--> Derive-Secret(., "astree-secret", ASTreeContext[V.leftChild])
+        |    = astree_node_[IndexOf(V.leftChild)]_secret
+        |
+        V
+Derive-Secret(., "astree-secret", ASTreeContext[V.rightChild])
+= astree_node_[IndexOf(V.rightChild)]_secret
+~~~~
 
-- Receivers MUST delete an Application Secret once it has been used to
-  derive the corresponding AEAD key and nonce as well as the next Application
-  Secret. Receivers MAY keep the AEAD key and nonce around for some
-  reasonable period.
+Note that fixing concrete values for GroupState_[n] and application_secret
+completely defines all secrets in the AS Tree.
 
-- Receivers MUST delete AEAD keys and nonces once they have been used to
-  successfully decrypt a message.
+## Sender Ratchets {#sender-ratchets}
 
-### Application AEAD Key Calculation
+The secret of a leaf in the AS Tree is used to initiate a symmetric hash ratchet
+which generates a sequence of keys and nonces. The group member assigned to that
+leaf uses the j-th key/nonce pair in the sequence to encrypt (using the AEAD)
+the j-th message they send during that epoch. In particular, each key/nonce pair
+MUST NOT be used to encrypt more than one message.
 
-The Application AEAD keying material is generated from the following
-input values:
+More precisely, the initial secret of the ratchet for the group member assigned to the leaf with index i is simply the secret of that leaf.
 
-- The Application Secret value;
-- A purpose value indicating the specific value being generated;
-- The length of the key being generated.
+~~~~
+application_[i]_[0]_secret = astree_node_[i]_secret
+~~~~
 
-Note, that because the identity of the participant using the keys to send data
-is included in the initial Application Secret, all successive updates to the
-Application secret will implicitly inherit this ownership.
+Keys, nonces and secrets of ratchets are derived using HKDF-Expand-Label. The
+context in the call is (a hash of) the current Group state, the index of sender's leaf and the position in that sender's ratchet.
 
-All the traffic keying material is recomputed whenever the underlying
-Application Secret changes.
+~~~~
+struct {
+  opaque gshash<0..2^32-1> = Hash(GroupState_[n]);
+  opaque leaf_index<0..2^16-1>;
+  uint32 ratchet_position;
+} RatchetContext
+~~~~
```

Comment:
My reasoning here is the same as I explained above: no 2 keys should have the same relation -> defence in depth. 

IMO TLS is designed for greater packets per second than MLS (and probably aims to run on weaker devices like IoT stuff too) so it makes sense for TLS to care a lot about being efficient. MLS though is meant for lower packet rate (on at least a cellphone level device) so I think its worth spending a small number of extra cycles and memory access for reasons of best practice.

Just as before, I'll leave it up to you which way to go though.

Reply:
My reasoning here is the same as I explained above: no 2 keys should have the same relation -> defence in depth.

IMO TLS is designed for greater packets per second than MLS (and probably aims to run on weaker devices like IoT stuff too) so it makes sense for TLS to care a lot about being efficient. MLS though is meant for lower packet rate (on at least a cellphone level device) so I think its worth spending a small number of extra cycles and memory access for reasons of best practice.

Just as before, I'll leave it up to you which way to go though.

Reply:
I mean leaf_index to denote the number of the leaf assigned to the owner of the ratchet. since the AS Tree mirrors the Ratchet Tree that's also the number of the leaf to which the ratchet is attached in the AS Tree.

All I really care about that the value is unique to that ratchet and that its easy to determine the number of any client in the group. So using a full identity seems like an over kill. The leaf_index should be enough.

Reply:
FYI I also removed the hashing of HashRatCont[i,j]. Now it just goes in the context as is instead of being hashed first.